### PR TITLE
make custom_transforms handle pytrees, add api.defvjp

### LINKED ISSUE #116: Easy api for custom primitives and vjps
JAX supports custom primitives and vjps, just like Autograd did. Improvements:
1) add this to documentation
2) add a minimal example of this in the examples section
3) add a wrapper function if appropriate?

### PR DESCRIPTION
**Pull request by mattjj**
Before this PR, using `custom_transforms` (from #419) with functions that took tuples/lists/dicts (i.e. pytrees) as arguments would silently fail. Moreover, defining custom rules that took as input or returned as output any pytree values wouldn't work. Finally, `custom_transforms` functions didn't automatically support vmap.

This PR fixes all that!

This PR also improves the API by adding `defjvp` and `defvjp` (cf. #636) functions (and their variants) to api.py, which can be called directly on a `custom_transforms` function rather than having to dig out a reference to the underlying primitive. There's also a `custom_gradient` convenience wrapper, which works like `tf.custom_gradient`.

This API will likely be the main mechanism by which users will define custom gradients.

TODO:
- [x] add docstrings

In the future we should add a tutorial explanation of how to do this in an .md file, like [in Autograd's tutorial](https://github.com/HIPS/autograd/blob/master/docs/tutorial.md#extend-autograd-by-defining-your-own-primitives), at which point we can finally close #116.

It would also be good to add more tests, especially for error messages.

### CODE REVIEW

**Comment by fehiepsi** [comment]
Hi @mattjj , will this support batching rule for transforms with 2 arguments? Currently, we have to use the following pattern for a custom transform with 2 arguments
```
def f_batching_rule(batched_args, batch_dims):
    x, y = batched_args
    bx, by = batch_dims
    # promote shapes
    sx, sy = np.shape(x), np.shape(y)
    nx = len(sx) + int(bx is None)
    ny = len(sy) + int(by is None)
    nd = max(nx, ny)
    x = np.reshape(x, (1,) * (nd - len(sx)) + sx)
    y = np.reshape(y, (1,) * (nd - len(sy)) + sy)
    # correct bx, by due to promoting
    bx = bx + nd - len(sx) if bx is not None else nd - len(sx) - 1
    by = by + nd - len(sy) if by is not None else nd - len(sy) - 1
    # move bx, by to front
    x = batching.move_dim_to_front(x, bx)
    y = batching.move_dim_to_front(y, by)
    return f(x, y), 0
```

**Comment by mattjj** [comment]
It should handle arbitrary numbers of arguments according to standard vmap semantics. Do you have a test case I can try?

**Comment by fehiepsi** [comment]
Sure, for example,
```
import jax.numpy as np
from jax import custom_transforms, vmap

@custom_transforms
def xlogy(x, y):
    return x * np.log(y)

vmap(xlogy)(np.ones(3), np.ones((3, 2)))
```

**Comment by mattjj** [comment]
That seems to work on this branch: it produces a zeros array of shape (3, 2). Is that the result you expect?

**Comment by fehiepsi** [comment]
Yup, that is the expected result. It is great to see that vmap is working for these custom transforms now.

File: jax/api.py [hunk 1]
```diff
@@ -952,25 +954,608 @@ def _valid_jaxtype(arg):
...
+  that specifies how each primitive is transformed. When a new primitive is
+  added to the system, it also needs a rule for each transformation we might
+  want to apply to it.
+
+  But in some cases instead of introducing a new primitive and specifying every
+  transformation rule, we might start with a traceable/transformable Python
+  function and want to override just its VJP (i.e. give it a custom gradient).
+  More concretely, we might start with a Python function to which we can already
+  apply ``jax.jit``, but we want to change how ``jax.grad`` behaves on it. These
+  are the cases that ``custom_transforms`` handles.
```
Comment (skye) [inline jax/api.py]: I'm not sure the above two paragraphs are necessary. They're true and might be useful general info for understanding how jax works, but I don't think are needed to understand how to use this custom_transforms. This docstring is already quite long so I would save these paragraphs for a future in-depth guide :)

File: jax/api.py [hunk 2]
```diff
@@ -952,25 +954,608 @@ def _valid_jaxtype(arg):
...
+
+  The ``custom_transforms`` decorator wraps ``fun`` so that its transformation
+  behavior can be overridden, but not all transformation rules need to be
+  specified manually. For rules that aren't overridden, a default rule is used
+  for each transformation that relies on tracing ``fun``.
```
Comment (skye) [inline jax/api.py]: Maybe replace last sentence with "The default behavior is retained for any non-overridden rules."

File: jax/api.py [hunk 3]
```diff
@@ -952,25 +954,608 @@ def _valid_jaxtype(arg):
...
+  if type(fun) is not CustomTransformsFunction:
+    msg = ("{} requires a custom_transforms function as its first argument, "
+          "but got type {}.")
+    raise TypeError(msg.format(name, type(fun)))
+
+def defjvp_all(fun, custom_jvp):
+  """Define a custom JVP rule for a ``custom_transforms`` function.
+
+  If ``fun`` represents a function with signature ``a -> b``, then
+  ``custom_jvp`` represents a function with signature ``a -> T a -> (b, Tb)``,
```
Comment (skye) [inline jax/api.py]: I still don't think the Haskell notation will be comprehensible to most users, but maybe it's part of learning jax :) At least do `T a` + `T b` (or `Ta` + `Tb`) to make the tangent notation consistent.

Edit: I see you explain it in words below, PHEW

File: jax/api.py [hunk 4]
```diff
@@ -952,25 +954,608 @@ def _valid_jaxtype(arg):
...
+  where we use ``T x`` to represent a tangent type for the type ``x``.
+
+  In more detail, ``custom_jvp`` must two arguments, both tuples of length equal
```
Comment (skye) [inline jax/api.py]: must take two

File: jax/api.py [hunk 5]
```diff
@@ -952,25 +954,608 @@ def _valid_jaxtype(arg):
...
-    # TODO(mattjj): pytrees to jaxtupletrees
-    return fun_p.bind(*args, **kwargs)
-  traceable.primitive = fun_p
+  Args:
+    fun: a custom_transforms function.
+    custom_jvp: a Python callable specifying the JVP rule, taking two tuples as
+      arguments specifying the input primal values and tangent values,
+      respectively. The tuple elements can be arrays, scalars, or (nested)
+      standard Python containers (tuple/list/dict) thereof. Must be functionally
+      pure.
```
Comment (skye) [inline jax/api.py]: Can the output be a pytree? Also briefly mention what the output is here.

File: jax/api.py [hunk 6]
```diff
@@ -952,25 +954,608 @@ def _valid_jaxtype(arg):
...
+    return ans, reduce(ad.add_tangents, tangents_out, ad_util.zero)
+  defjvp_all(fun, custom_jvp)
+
+def defjvp2(fun, *jvprules):
+  """Definine JVP rules for each argument separately.
+
+  This function is a convenience wrapper around ``jax.defjvp_all`` for separately
+  defining JVP rules for each of the function's arguments. This convenience
+  wrapper does not provide a mechanism for depending on anything other than the
+  function arguments and its primal output value.
```
Comment (skye) [inline jax/api.py]: Maybe mention here or elsewhere that it's like defjvp, except the jvp function also has access the primal output.

File: jax/api.py [hunk 7]
```diff
@@ -952,25 +954,608 @@ def _valid_jaxtype(arg):
...
+  """
+  _check_custom_transforms_type("defjvp", fun)
+  def custom_jvp(primals, tangents):
+    ans = fun(*primals)
+    tangents_out = [rule(t, *primals) for rule, t in zip(jvprules, tangents)
+                    if rule is not None and t is not ad_util.zero]
+    return ans, reduce(ad.add_tangents, tangents_out, ad_util.zero)
+  defjvp_all(fun, custom_jvp)
+
+def defjvp2(fun, *jvprules):
```
Comment (skye) [inline jax/api.py]: I'm not sure what 2 refers to, maybe the two non-variadic arguments to the jvprules functions? Could we get rid of defjvp and only have this one?

**Review by skye — COMMENTED** [review]
Still reviewing, gonna send these out in case you wanna pipeline reviewing + responding...

File: jax/api.py (hunk 1 above)
Comment (mattjj) [reply jax/api.py]: Done.

File: jax/api.py (hunk 3 above)
Comment (mattjj) [reply jax/api.py]: Done.

File: jax/api.py (hunk 4 above)
Comment (mattjj) [reply jax/api.py]: Done.

File: jax/api.py (hunk 5 above)
Comment (mattjj) [reply jax/api.py]: Done.

File: jax/api.py (hunk 6 above)
Comment (mattjj) [reply jax/api.py]: Done.

File: jax/api.py (hunk 7 above)
Comment (mattjj) [reply jax/api.py]: We could just have one, but the most common pattern in our internal code (where we have `ad.defjvp` and `ad.defjvp2` for Primitives) is not to depend on the output. That is, if you look at our JVP rules in lax.py, we almost always use ``defejvp`` but then sometimes use ``defjvp2``. That and having these functions follow the convention of their counterparts in ad.py seem like good reasons.

The 2 doesn't refer to anything, it's just a way to name a variant of a function while keeping the name short. To me it's like a "prime", but we can't add `'` to a name in Python.

I know you might not like that convention, but we already use it elsewhere (to no ill effect so far!), and in any case it doesn't seem like something worth getting stuck on. WDYT?

File: jax/api.py (hunk 2 above)
Comment (mattjj) [reply jax/api.py]: Done.

File: jax/interpreters/ad.py [hunk 8]
```diff
@@ -397,18 +397,19 @@ def add_tangents(x, y):
 
 
 def defvjp_all(prim, custom_vjp):
```
Comment (skye) [inline jax/interpreters/ad.py]: For my understanding, how is this related to the defvjp_all in api.py?

File: jax/api.py [hunk 9]
```diff
@@ -952,25 +954,608 @@ def _valid_jaxtype(arg):
...
+                         for x, vjp in zip(primals, vjprules)]
+    return ans, vjpfun
+  defvjp_all(fun, custom_vjp)
+
+def custom_gradient(fun):
+  """Convenience function for defining custom VJP rules (aka custom gradients).
+
+  While the canonical way to define custom VJP rules is via ``jax.defvjp_all``
+  and its convenience wrappers, the ``custom_gradient`` convenience wrapper
+  follows TensorFlow's tf.custom_gradient API.
```
Comment (skye) [inline jax/api.py]: Can you add a brief description of how this is different from defvjp_all, and/or when you would use one or the other? IIUC, custom_gradient is not used with custom_transforms, and you define the "forward-pass" (primal?) function and the VJP in a single function `fun`, and then the returned function is just the forward pass. So I guess you use this when you wanna quickly define a forward and backward computation in one shot? But isn't as useful for applying a VJP rule to an already-existing function? (If using one vs the other is really a matter of taste, no need to include that part.)

File: jax/api.py [hunk 10]
```diff
@@ -952,25 +954,608 @@ def _valid_jaxtype(arg):
...
+
+  See https://www.tensorflow.org/api_docs/python/tf/custom_gradient.
+
+  If the function to be differentiated has type signature ``a -> b``, then the
```
Comment (skye) [inline jax/api.py]: "the function to be differentiated" is kind of confusing because there's not necessarily a Python function corresponding to it (i.e. it's not `fun`). Maybe... "the primal computation"? Not sure if I'm using these terms correctly...

File: jax/api.py (hunk 7 above)
Comment (shoyer) [reply jax/api.py]: Maybe `defjvp_separate` or  `devjvp_sep` if you really want to save characters?

For what it's worth I also found the name confusing, and wondered what the 2 was about :)

File: jax/api.py [hunk 11]
```diff
@@ -952,25 +954,597 @@ def _valid_jaxtype(arg):
...
+  Python callable ``fun`` should have signature ``a -> (b, CT b -> CT a)`` where
+  we use ``CT x`` to denote a cotangent type for ``x``. See the example below.
+  That is, ``fun`` should return a pair where the first element represents the
+  value of the function to be differentiated and the second element is a
+  function that represents the custom VJP rule.
+
+  The custom VJP function returned as the second element of the output of ``fun``
+  can close over intermediate values computed when evaluating the function to be
+  differentiated. That is, use lexical closure to share work between the forward
+  pass and the backward pass of reverse-mode automatic differentiation.
```
Comment (shoyer) [inline jax/api.py]: It would be nice to show an example of this, e.g., copied from tf.custom_gradient.

File: tests/api_test.py [hunk 12]
```diff
@@ -507,20 +507,30 @@ def vjpfun(x, y):
...
     val_ans, grad_ans = api.value_and_grad(foo)(3., 4.)
     self.assertAllClose(val_ans, 3.**2 + 4.**3, check_dtypes=False)
     self.assertAllClose(grad_ans, 1. + 3. + 4., check_dtypes=False)
 
     ans = api.grad(foo, (0, 1))(3., 4.)
     self.assertAllClose(ans, (1. + 3. + 4., 1. * 3. * 9.), check_dtypes=False)
 
+  def test_defvjp_all(self):
+    @api.custom_transforms
+    def foo(x):
```
Comment (skye) [inline tests/api_test.py]: Maybe add a test for a function that takes multiple and/or pytree arguments? (Sorry if I missed it below)

**Review by skye — COMMENTED** [review]
Finished reviewing the original PR, gonna look at changes now...

File: jax/api.py (hunk 7 above)
Comment (fehiepsi) [reply jax/api.py]: I vote for the current name: `defjvp2` by the same reason as Matt.

Comment (skye) [reply jax/api.py]: For jvp rules that don't depend on the output, can't you just put a `_` for that arg and ignore it? If the performance etc. is the same, seems pretty lightweight to me. People (including us) can make their own wrappers if it really bothers them.

Comment (mattjj) [reply jax/api.py]: > For what it's worth I also found the name confusing, and wondered what the 2 was about :)

Wondering what a function is and how it works doesn't seem like a bad thing! I would wonder about what all of these functions mean before looking at example uses or their code (or docstring). ("separate" isn't the right suffix here because the difference in question is between taking the output primal as an argument or not.)

> People (including us) can make their own wrappers if it really bothers them.

Yes, that's the idea, but since `defjvp` is the most common pattern, and `defjvp2` is the more general second-most-common pattern, we provide those wrappers for users. After all, both `defjvp` and `defjvp2` are already convenience wrappers around `defjvp_all`.

I think reasonable people can (and do) disagree about personal preferences here. I'm going to keep `defjvp2`.

File: jax/interpreters/ad.py (hunk 8 above)
Comment (mattjj) [reply jax/interpreters/ad.py]: This one applies to primitives (i.e. its first argument must be a core.Primitive instance), while the one in api.py applies to `custom_transforms` functions. (Those two paragraphs that we deleted from the `custom_transforms` docstring explains how a Primitive is different from a custom_transforms function!)

File: jax/api.py (hunk 11 above)
Comment (mattjj) [reply jax/api.py]: Thanks for the suggestion! The second example in the docstring does use lexical closure, though I agree more (and more embellished) examples can help. I think the best place for them would be an .md file explaining how to use these, rather than the docstring.

File: jax/api.py (hunk 10 above)
Comment (mattjj) [reply jax/api.py]: There's not a Python function (i.e. a Python callable) but there is a mathematical function. I agree the phrasing seems tough but I'm not sure "the primal computation" is any better than "the function to be differentiated"...

File: jax/interpreters/ad.py (hunk 8 above)
Comment (skye) [reply jax/interpreters/ad.py]: Oh right. I think I figured that out but forgot to delete this review comment, my bad!

File: jax/api.py (hunk 10 above)
Comment (skye) [reply jax/api.py]: Maybe just "the mathematical function to be differentiated"?

File: tests/api_test.py (hunk 12 above)
Comment (skye) [reply tests/api_test.py]: (This can be a follow-up change, but I think is worth doing, at the very least to make sure we have a runnable example of multiple args.)

File: jax/api.py (hunk 9 above)
Comment (mattjj) [reply jax/api.py]: Done.

File: jax/api.py (hunk 10 above)
Comment (mattjj) [reply jax/api.py]: Done.

File: tests/api_test.py (hunk 12 above)
Comment (mattjj) [reply tests/api_test.py]: Will do in a follow-up; left a todo.

**Comment by mattjj** [comment]
@skye convinced me she was right about not having `api.defjvp2`/`api.defvjp2` on API minimalism grounds (rather than readability grounds). I updated the PR to only have a `defjvp` and `defvjp`, though both act like the "2" variants in that they take `ans` as an argument (which can be ignored).
